src/main/java/com/spotify/annoy/ANNIndex.java

```diff
@@ -66,46 +57,98 @@ public ANNIndex(final int dimension,
     this(dimension, filename, IndexType.ANGULAR);
   }
 
+  ANNIndex(final int dimension,
+                  final String filename,
+                  IndexType indexType,
+                  final int blockSize) throws IOException {
+    DIMENSION = dimension;
+    INDEX_TYPE = indexType;
+    INDEX_TYPE_OFFSET = (INDEX_TYPE == IndexType.ANGULAR) ? 4 : 8;
+    K_NODE_HEADER_STYLE = (INDEX_TYPE == IndexType.ANGULAR) ? 12 : 16;
+    // we can store up to MIN_LEAF_SIZE children in leaf nodes (we put
+    // them where the separating plane normally goes)
+    this.MIN_LEAF_SIZE = DIMENSION + 2;
+    this.NODE_SIZE = K_NODE_HEADER_STYLE + FLOAT_SIZE * DIMENSION;
+    this.BLOCK_SIZE = blockSize == 0 ?
+            Integer.MAX_VALUE / NODE_SIZE : blockSize * NODE_SIZE;
+
+    roots = new ArrayList<>();
+    load(filename);
+  }
 
   private void load(final String filename) throws IOException {
     memoryMappedFile = new RandomAccessFile(filename, "r");
-    int fileSize = (int) memoryMappedFile.length();
-    // We only support indexes <4GB as a result of ByteBuffer using an int index
-    annBuf = memoryMappedFile.getChannel().map(
-            FileChannel.MapMode.READ_ONLY, 0, fileSize);
-    annBuf.order(ByteOrder.LITTLE_ENDIAN);
+    long fileSize = memoryMappedFile.length();
+    if (fileSize == 0L) {
+      throw new IOException("Index is a 0-byte file?");
+    }
+
+    int buffIndex =  (int) (fileSize - 1) / BLOCK_SIZE;
+    int rest = (int) (fileSize % BLOCK_SIZE);
+    int blockSize = (rest > 0 ? rest : BLOCK_SIZE);
+    long position = Math.max(0, fileSize - blockSize);
+
+    buffers = new MappedByteBuffer[buffIndex + 1];
+    boolean process = true;
     int m = -1;
-    for (int i = fileSize - NODE_SIZE; i >= 0; i -= NODE_SIZE) {
-      int k = annBuf.getInt(i);  // node[i].n_descendants
-      if (m == -1 || k == m) {
-        roots.add(i);
-        m = k;
-      } else {
-        break;
+    long index = fileSize;
+    while(position >= 0) {
+      MappedByteBuffer annBuf = memoryMappedFile.getChannel().map(
+              FileChannel.MapMode.READ_ONLY, position, blockSize);
+      annBuf.order(ByteOrder.LITTLE_ENDIAN);
+
+      buffers[buffIndex--] = annBuf;
+
+      for (int i = blockSize - NODE_SIZE; process && i >= 0; i -= NODE_SIZE) {
+        index -= NODE_SIZE;
+        int k = annBuf.getInt(i);  // node[i].n_descendants
+        if (m == -1 || k == m) {
+          roots.add(index);
+          m = k;
+        } else {
+          process = false;
+        }
       }
+      blockSize = BLOCK_SIZE;
+      position = position - blockSize;
```

Comment:
Ok, sorry, I was confused. Because it's being read backwards, you're right.
```
|  BLOCK 1  |  BLOCK 2  |  BLOCK 3 |  BLN |
                                   ^ position is initially here
                                   -------- blockSize is initially this big
```
obviously you need to decrement position by `BLOCK_SIZE`, not `blockSize` here.